leech/src/modules/service_detection/udp.rs

```diff
@@ -0,0 +1,169 @@
+use std::net::{IpAddr, SocketAddr};
+use std::ops::RangeInclusive;
+use std::time::Duration;
+
+use futures::{stream, TryStreamExt};
+use log::{debug, info};
+use probe_config::generated::Match;
+use tokio::net::UdpSocket;
+use tokio::sync::mpsc::Sender;
+use tokio::task::JoinSet;
+use tokio::time::{sleep, timeout};
+
+use super::error::UdpServiceScanError;
+use super::Service;
+use crate::modules::service_detection::generated;
+
+/// Settings for a service detection
+#[derive(Clone)]
+pub struct UdpServiceDetectionSettings {
+    /// IP address to scan
+    pub ip: IpAddr,
+
+    /// Ports to scan
+    pub ports: Vec<RangeInclusive<u16>>,
+
+    /// Number of times packet sending should be retried if no response was received within the `timeout`
+    ///
+    /// Set to 0 to only try once.
+    pub max_retries: u32,
+
+    /// How much time to wait between sending probe payloads (independent of
+    /// timeout, aka it can send while another pending probe did not yet receive
+    /// data within the timeout interval).
+    pub retry_interval: Duration,
+
+    /// The timeout for receiving data on a single sent probe.
+    pub timeout: Duration,
+
+    /// Maximum of concurrent tasks that should be spawned
+    ///
+    /// 0 means, that there should be no limit.
+    pub concurrent_limit: u32,
+}
+
+impl UdpServiceDetectionSettings {
+    async fn probe_udp(
+        &self,
+        port: u16,
+        payload: &[u8],
+    ) -> Result<Option<Vec<u8>>, UdpServiceScanError> {
+        let addr = SocketAddr::new(self.ip, port);
+
+        let mut set = JoinSet::new();
+
+        for i in 0..=self.max_retries {
+            let offset = self.retry_interval.saturating_mul(i);
+            set.spawn(timeout(
+                self.timeout.saturating_add(offset),
+                single_probe_udp(offset, addr, payload.to_vec()),
+            ));
+        }
+
+        while let Some(res) = set.join_next().await {
+            let res = res?; // propagate JoinError
+            if let Ok(res) = res {
+                // ignore timeouts
+                let res = res?; // propagate single_probe_udp errors
+                if let Some(res) = res {
+                    // check for single_probe_udp returning None (aka no response)
+                    set.abort_all();
+                    return Ok(Some(res));
+                }
+            } else {
+                debug!("timeout on udp service probe {}", addr);
+            }
+        }
+
+        Ok(None)
+    }
+}
+
+/// Returns the received payload or None in case there was no response. Returns
+/// errors in case of other networking errors.
+async fn single_probe_udp(
+    wait_before: Duration,
+    addr: SocketAddr,
+    payload: Vec<u8>,
+) -> Result<Option<Vec<u8>>, UdpServiceScanError> {
+    sleep(wait_before).await;
+    let sock = UdpSocket::bind("0.0.0.0:0").await?;
```

Comment:
Also consider ipv6

Reply:
implemented, although was unable to confirm full functionality